Leetcode/aoc_2015_d4.py

```diff
@@ -0,0 +1,140 @@
+
+'''
+ ----- Thoughts on part 1 after completing it. -----
+This one was a bit tricky. I understood it conceptually, but I didn't know
+how to put it in code as easily since I didn't know how the hashlib and md5
+worked. It took a bit of trial and error, but understanding it conceptually
+made coding it a bit easier, I think.
+Since it must go through numbers by iteration, it can take a good amount of time
+find the right string that leads with 5 zeroes.
+If I really wanted to improvet this, I could probably do something similar to my
+java mega millions program and have it multithreaded with each loop going through
+say 999_999 numbers. e.x. thread 1 does 1 to 100_000, thread 2 does 1_000_000 to
+1_999_999 numbers. But that's micro-optimization, since there's no guarantee it'd
+be a number that high. For example, my secret key had an answer that was quite low.
+'''
+
+'''
+ ----- Part 2 -----
+Part 2 is the same thing, just finding one that starts with six zeroes
+instead of five. Seeing this, I can now see that my thought on multi-threading
+would be more helpful as finding one with six leading zeroes would take much
+more time.
+Part 2 would be the same code as part 1, just changing the while conditional.
+So to take it a step further, I'll just add another parameter to the function.
+and set the while loop to use that.
+
+Knowing that this will take longer, I have decided to go back in both
+part 2 and part 1 to get execution time.
+This is done by getting cpu time using the time library and using process_time()
+'''
+
+# This function can completely replace the part 1 version.
+# Works the exact same way as the first one, default is 5 leading zeroes.
+# can just do find_hash_p2(h) and it'll look for 5 leading,
+# but you can pass it a different amount of leading zeroes to look for
+# and it will look for that instead.
+def find_hash_p2(h: str, leading=5) -> int:
+	target_hash = "0"
+	find_hash = 0
+	# another adjustment, since it's now a different amount
+	# of leading zeroes, we're going to slice off the amount
+	# of leading we want instead of a set 5.
+	# then we're going to multiply the 0s by the amount of leading
+	# we want.
+	# Breaks if leading is passed a 0.
+	while target_hash[:leading] != leading * "0":
+		find_hash += 1
+		str2hash = (h + str(find_hash)).encode()
+		target_hash = hashlib.md5(str2hash).hexdigest()
+
+	return find_hash
+
+st = time.process_time()
+print(find_hash(h))
+et = time.process_time()
+print(f'5 Leading zeroes\n')
+print(f'Execution Time: {et - st} seconds\n')
+# My secret key took 0.25 seconds.
+
+st = time.process_time()
+print(find_hash_p2(h, 6))
+et = time.process_time()
+print(f'6 Leading zeroes\n')
+print(f'Execution Time: {et - st} seconds\n')
+# My secret key took 1.01 seconds.
+
+# Aaaand after running it, maybe you don't need to multi-thread it,
+# it's quite fast anyway.
+
+# Not entirely sure how to figure out the time complexity of this,
+# since the number could theoretically be infinite(?).
```

Comment:
You're mostly correct. Analyzing the time complexity of this is largely nonsensical - but consider that MD5 is a constant time algorithm (the inputs don't make it run faster or slower), this is O(n) linear in terms of the nonce counter.

Infinite is not quite correct, because as noted in the above ChatGPT scenario, there's only a finite number of possible hashes, since hashes always result in a 32 character hexadecimal string. In fact, it's this cap on the available hashes that causes hash collisions (two different pieces of input that end up hashing to the same thing).

Reply:
I see, that makes sense.